### PR TITLE
Moving `Account.amount` to `Account.credit` & `Account.debit`

### PR DESCRIPTION
**Pull request by adamcharnock**
I've had a go at this this morning, and you can see the [the core changes here](https://github.com/adamcharnock/django-hordak/pull/133/commits/2970b151c2d6f3a55e9922023e3c8f2b7332c89e#diff-8c8b13fb152bf1f22728c2cb7742637f15f435a1247c9efb73b9208fc1bbe0a5).

Most of the lines changed are tests.

# Notes:

* It is a straightforward but far-reaching change
* I was hoping this would drop out a bunch of complexity, but it did not
* `my_leg.amount` is still available as a property on the Django model. It is now always a positive value.
* Can still use the `amount` argument when creating a Leg, a deprication warning will be shown.
* I've added `amount_legacy` (I.e. debits are negative) to the LegView, so it is still available at `my_leg.view.amount_legacy`
* I'm not against merging this for Hordak 2.0 or 3.0
* **The more time I spend with this branch the more it does feel like a much better way to structure things. I'm rather in favour of merging this.**

# Remaining Tasks

- [x] Test against MySQL
- [x] Trigger to ensure that only credit or debit is set
  - [x] Test
- [x] Trigger to ensure that credit/debit is > 0
  - [x] Test

# Follow-up work

I'll do this next in another PR

- Ensure 'balance()' functionality hasn't changed (is the sign of the return value still the same?)
- Implement 'accounting_balance()' if necessary 
- Test data migration (related: #111)

### CODE REVIEW

**Comment by adamcharnock** [comment]
All tests now passing 🎉🎉🎉

File: hordak/models/core.py [hunk 1]
```diff
@@ -520,29 +542,80 @@ class Leg(models.Model):
...
-        return f"{self.type.title()} {self.account.name} ({self.account.full_code}) {self.amount}"
+        return (
+            f"{self.type.title()} {self.account.name} "
+            f"({self.account.full_code}) {self.amount} {self.type_short}"
+        )
+
+    def __init__(self, *args, amount: Money = None, **kwargs):
+        if amount is not None:
+            warnings.warn(
+                "Specifying `amount` when creating an Account is deprecated. "
```
Comment (nitsujri) [inline hordak/models/core.py:553]: I think this is supposed to say `... when creating a Leg...` not Account?

File: hordak/tests/test_worked_examples.py [hunk 2]
```diff
@@ -155,7 +155,7 @@ def test_utility_bill(self):
 
         # We receive the actual bill (we are moving a negative amount of money,
         # as this is an outgoing)
-        self.gas_payable.transfer_to(self.cash, Money(-300, "EUR"))
+        self.cash.transfer_to(self.gas_payable, Money(300, "EUR"))
```
Comment (nitsujri) [inline hordak/tests/test_worked_examples.py:158]: This is incredible! I'm insanely excited to see this change right here.

I'm sure you already know this, but just stating it explicitly to be on the same page: 

This is a breaking change for anyone doing RHS -> LHS, vice versa - using the current `.transfer_to`. Saw that `accounting_transfer_to` "effectively" replaces it which I think is 100% correct, just it's a significant change and warrants an "Upgrade to 2.0" document.

**Review by nitsujri — COMMENTED** [review]
Love everything about this. I won't be able to test this until August due to travels and deadlines, but in spirit, I 100% approve and happy to see it go out ahead of our team's direct testing.

File: hordak/models/core.py (hunk 1 above)
Comment (adamcharnock) [reply hordak/models/core.py]: Ooh yes, great catch! Fixed.

File: hordak/tests/test_worked_examples.py (hunk 2 above)
Comment (adamcharnock) [reply hordak/tests/test_worked_examples.py:158]: I'm very pleased to hear it! And yes, absolutely understood re upgrading. I've [put these notes in the changelog](https://github.com/adamcharnock/django-hordak/blob/experiment/amount-to-debit-credit/CHANGES.txt). How does that seem to you?

File: CHANGES.txt [hunk 3]
```diff
@@ -1,9 +1,19 @@
 Version 2.0.0 (unreleased)
 ==========================
 
-* **Breaking**: New account currencies now default to ``DEFAULT_CURRENCY`` rather than all available currencies (trading accounts should be the only accounts with multiple currencies).
+**Important:** It is strongly recommended that you
+**upgrade to Hordak 1.16 before upgrading to Hordak 2.0**. Update your code
+to address any depreciation warnings in 1.16 before upgrading to Hordak 2.0.
+
+* **Breaking**: ``to_account()`` was deprecated in 1.16. This old implementation has now been removed
```
Comment (nitsujri) [inline CHANGES.txt:8]: I think you mean `transfer_to()` and `accounting_transfer_to()`

Comment (adamcharnock) [reply CHANGES.txt]: Ah yes, thank you @nitsujri! Fixed.

**Comment by adamcharnock** [comment]
Hi @PetrDlouhy – I'm inclined to merge this sometime soon. If you wanted to have a dramatic "I object!" moment then please do so soon (other reasoned input also very welcome)

**Comment by nitsujri** [comment]
Would it be possible to remove the views? I believe the views are now no longer needed?

Specifically migrations against any column that is part of a view becomes a pain.

**Comment by PetrDlouhy** [comment]
Ať first glance this seems reasonable. I don't have any direct objection. 
I would have to find some time to test this with my project to tell more.

File: hordak/admin.py [hunk 4]
```diff
@@ -51,8 +52,12 @@ def get_queryset(self, *args, **kwargs):
             super()
             .get_queryset(*args, **kwargs)
             .annotate(
-                balance_sum=Sum("legs__amount"),
-                income=Sum("legs__amount", filter=Q(legs__amount__gt=0)),
+                # TODO: I think this needs some sign-changing work
```
Comment (nitsujri) [inline hordak/admin.py:55]: Agreed, something like `if obj.type` then flip the subtraction?

**Review by nitsujri — COMMENTED** [review]
Was able to test it on our codebase. Overall looks great after internal modifications.

Recommended changes:

- Remove views
- Leg.credit and debit `max_digits` needs to use `HORDAK_MAX_DIGITS`.
- `get_queryset` needs to swap signs based on account type.

File: hordak/migrations/0044_legview.py [hunk 5]
```diff
@@ -368,7 +370,7 @@ class Migration(migrations.Migration):
                 (
                     "credit",
                     models.DecimalField(
-                        decimal_places=6,
+                        decimal_places=DECIMAL_PLACES,
                         help_text="Amount of this credit, or NULL if not a credit",
                         max_digits=20,
```
[the remark this answers is not in the document]
Comment (adamcharnock) [reply hordak/migrations/0044_legview.py:375]: Thanks @nitsujri, I saw this one too. I've just double checked and a fix for this is already committed.

File: hordak/admin.py (hunk 4 above)
Comment (adamcharnock) [reply hordak/admin.py]: Done

**Comment by adamcharnock** [comment]
@nitsujri 

> Leg.credit and debit max_digits needs to use HORDAK_MAX_DIGITS.

Done

> get_queryset needs to swap signs based on account type.

Done

> Remove views

Done

I'm a bit loathed to removed them as I think they may be useful, but I'm also willing to be found wrong on this. I've therefore added this warning to the docs for each view: _Hordak's database views are still experimental and may change or be removed in a future version._

In the mean-time, my hope is that the new way of structuring sql-heavy migrations will make this problem less onerous.

**Comment by adamcharnock** [comment]
I've also made some additional changes, in particular moving many balance calculations into annotations. I think this should help performance a lot. Once this settles we can see if something like #76 (running totals) is still required.

I'll merge this soon unless anyone cries out. Let me know if you find any time for some testing @PetrDlouhy. I could always do an alpha-release if that helps at all.

File: hordak/migrations/0051_amount_to_debit_credit_data.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    operations = [
+        # On an M2 Macbook Pro, this took 16 seconds on 600k rows.
+        # Therefore on 10M rows this may take around 5 minutes.
+        migrations.RunSQL(
+            """
+                UPDATE
+                    hordak_leg
+                SET
+                    credit = CASE WHEN amount > 0 THEN amount END,
+                    debit = CASE WHEN amount < 0 THEN amount END,
```
Comment (nitsujri) [inline hordak/migrations/0051_amount_to_debit_credit_data.py:20]: Tried migrating our staging system in preps for production. Found the `debit` column was holding a negative value. Same with the views.

I believe it should be flipped?

`debit = CASE WHEN amount < 0 THEN -amount END,`

**Review by nitsujri — CHANGES_REQUESTED** [review]
`debit` column still holds negative value post migration - blocking issue.

File: hordak/migrations/0051_amount_to_debit_credit_data.py (hunk 6 above)
Comment (adamcharnock) [reply hordak/migrations/0051_amount_to_debit_credit_data.py]: Thank you for catching this! Fixed.

**Comment by adamcharnock** [comment]
A pre-release is now out as `2.0.0a1` 🎉🎉🎉🎉🎉

A huge thank you to @nitsujri and @PetrDlouhy for working through this with me.

**Pull request merged**